src/core/server/saved_objects/migrationsv2/model.ts

```diff
@@ -576,17 +579,68 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
           outdatedDocuments: res.right.outdatedDocuments,
         };
       } else {
-        // If there are no more results we have transformed all outdated
-        // documents and can proceed to the next step
+        if (stateP.failedDocumentIds && stateP.failedDocumentIds?.length > 0) {
+          // we exit out here and throw an error with the ids of documents that we'ren't transformed
+          throw new CorruptSavedObjectError(stateP.failedDocumentIds?.toString());
```

Comment:
nit: I've been trying to avoid using exceptions for control flow. So basically we only throw an exception if something unexpected happens that we have no idea how to handle. If there's a known failure condition we go to the `FATAL` state. E.g. when there's a newer Kibana we want to fail, but this is a condition we fully expect to happen so we handle it explicitly https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrationsv2/model.ts#L715-L716

Reply:
Ok, that makes sense, especially because we're handling the failed docs as a `right`. Should we then rather handle this as a failure condition? We want migrations to stop after we've built up the full list of failed docs. Should we rather try the approach of creating a `TaskEither` from the async transform method?

Reply:
Instead of throwing we would just go to the `FATAL` state with a `reason` string. The code changed after I posted, so the link is no longer showing the correct lines.

But I meant something like:
```
        return {
          ...stateP,
          controlState: 'FATAL',
          reason: `... migrations failed because of the following corrupt saved object documents: [...]`,
        };
```

Reply:
I thought about that too but then realized we're specifically [targeting](https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrationsv2/migrations_state_action_machine.ts#L180) `CorruptSavedObjectError` within the `catch` block of `migrationStateActionMachine`, where we throw a [`new Error`](https://github.com/elastic/kibana/blob/master/src/core/server/saved_objects/migrationsv2/migrations_state_action_machine.ts#L181-L183) with a custom message. I guess we can remove that.